[WIP][DI][for discussion] Allow explicit executor for torchscript interpreter

garroud opened the pull request:
Summary: add explicit executor for torchscript interpreter.

Test Plan:
unit test failed due to new added test in fbcode/caffe2/test/cpp/jit/test_interpreter.cpp

Seems due to failure of compiling graph with error msg:

```
attribute lookup is not defined on builtin:

```

Differential Revision: D24114874

dzhulgakov commented on torch/csrc/jit/runtime/interpreter.h:
```diff
@@ -66,7 +67,7 @@ struct TORCH_API Code {
 
 struct InterpreterState {
-  TORCH_API InterpreterState(const Code& code);
+  TORCH_API InterpreterState(const Code& code, TaskExecutor task_executor);
```
maybe provide default value here too? then you don't need to update all unittests

dzhulgakov commented on test/cpp/jit/test_interpreter.cpp:
```diff
@@ -131,12 +133,36 @@ TEST(InterpreterTest, Basic_CUDA) {
+  auto count = 0;
+  // a dummy executor
+  auto executor = [&](std::function<void()> f) {
+    ++count;
```
you probably do want to call `f()` here actually

dzhulgakov commented on test/cpp/jit/test_interpreter.cpp:
```diff
@@ -131,12 +133,36 @@ TEST(InterpreterTest, Basic_CUDA) {
+  // a simple tast executor
+  const auto moduleExample = R"JIT(
+  def run_parallel():
+    r1 = torch.jit.fork(torch.mm, torch.rand(10000,1000),torch.rand(1000,5000))
```
1000x5000 matrices might be a bit slow for a unittest

Commit:
[DI] Allow explicit executor for torchscript interpreter (#46362)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/46362

add explicit executor for torchscript interpreter.

Test Plan: unit test is passed.

Differential Revision: D24114874

fbshipit-source-id: 8f0f406fa81275a51b5f361fd9c6933be2af7258
